src/Microsoft.Identity.Web.TokenCache/MsalAbstractTokenCacheProvider.cs

```diff
@@ -98,7 +112,23 @@ private async Task OnBeforeAccessAsync(TokenCacheNotificationArgs args)
             {
                 byte[] tokenCacheBytes = await ReadCacheBytesAsync(args.SuggestedCacheKey, CreateHintsFromArgs(args)).ConfigureAwait(false);
 
-                args.TokenCache.DeserializeMsalV3(UnprotectBytes(tokenCacheBytes), shouldClearExistingCache: true);
+                try
+                {
+                    args.TokenCache.DeserializeMsalV3(UnprotectBytes(tokenCacheBytes), shouldClearExistingCache: true);
+                }
+                catch (Exception ex)
+                {
+                    // clears the cache
+                    args.TokenCache.DeserializeMsalV3(null, shouldClearExistingCache: true);
+                    if (_logger != null)
+                    {
+                        Logger.CacheDeserializationError(
```

Comment:
DataProtection API is a complex system that deals with encryption key rotation and key distribution among all the machines involved in the API. This is all taken care of for you, which is pretty amazing. Please see https://docs.microsoft.com/en-us/aspnet/core/security/data-protection/implementation/key-management?view=aspnetcore-6.0 for details.

But keys also need to be encrypted at rest. This is where it gets tricky.

The customer who logged this bug uses certificates for key encryption at rest: 

```csharp
builder.Services.AddDataProtection()
    .PersistKeysToBlobStorage(...)
    .ProtectKeysWithCertificate(cert)
```

This works, but when it comes time to rotate that cert, rotation needs to be handled explicitly by the app developer: 

```csharp
builder.Services.AddDataProtection()
    .PersistKeysToBlobStorage(...)
    .ProtectKeysWithCertificate(new_cert)
    .UnprotectKeysWithAnyCertificate(new_cert, old_cert);
```

This is described here: https://docs.microsoft.com/en-us/aspnet/core/security/data-protection/configuration/overview?view=aspnetcore-6.0#unprotectkeyswithanycertificate

Reply:
@pmaytak - pls ask the app developer who raised this issue to try this out.